### PR TITLE
Use a stream id generator to assign state group ids

### PR DESCRIPTION
**Pull request by NegativeMjark**

### CODE REVIEW

File: synapse/events/__init__.py [hunk 1]
```diff
@@ -31,7 +31,7 @@ def get_dict(self):
         return dict(self.__dict__)
 
     def is_outlier(self):
-        return hasattr(self, "outlier") and self.outlier
+        return getattr(self, "outlier", False)
```
Comment (NegativeMjark) [inline synapse/events/__init__.py]: This isn't strictly relevant to the PR, but I came across it when trying to see how we decide when to store a state group.

File: synapse/storage/events.py [hunk 2]
```diff
@@ -178,7 +195,7 @@ def get_events(self, event_ids, check_redacted=True,
 
     @log_function
     def _persist_event_txn(self, txn, event, context,
-                           is_new_state=True, current_state=None):
+                           is_new_state, current_state):
```
Comment (NegativeMjark) [inline synapse/storage/events.py]: Again, this wasn't strictly relevant to the PR, but there isn't much point in giving defaults in an internal function called from exactly one place.

File: synapse/storage/events.py [hunk 3]
```diff
@@ -310,19 +325,14 @@ def _persist_events_txn(self, txn, events_and_contexts, backfilled,
...
+            ec for ec in events_and_contexts if ec[0] not in to_remove
+        ]
 
         if not events_and_contexts:
             return
 
-        self._store_mult_state_groups_txn(txn, [
-            (event, context)
-            for event, context in events_and_contexts
-            if not event.internal_metadata.is_outlier()
```
Comment (NegativeMjark) [inline synapse/storage/events.py]: We can do the is_outlier check inside _store_mult_state_groups_txn.

**Comment by erikjohnston** [comment]
LGTM

**Pull request merged**